hg-fast-export.py

```diff
@@ -127,16 +127,20 @@ def get_author(logmessage,committer,authors):
       return r
   return committer
 
+def remove_gitmodules():
+  # Remove all submodules as we don't detect deleted submodules properly
+  # in any other way. We will add the ones not deleted back again below.
+  for module in submodule_mappings.keys():
+    wr('D %s' % module)
+  
+  wr('D .gitmodules')
+
 def export_file_contents(ctx,manifest,files,hgtags,encoding='',plugins={}):
   count=0
   max=len(files)
-  for file in files:
-    if submodule_mappings and ctx.substate and file==".hgsubstate":
-      # Remove all submodules as we don't detect deleted submodules properly
-      # in any other way. We will add the ones not deleted back again below.
-      for module in submodule_mappings.keys():
-        wr('D %s' % module)
 
+  for file in files:
+    if submodule_mappings and file==".hgsubstate":
```

Comment:
Can you explain why you drop `ctx.substate`?

Reply:
Ah, it's residual change from my first attempt to locate and fix this bug. Previously after removal of only subrepo it prevented update of `.gitmodules` file.

Now that isn't have a meaning, so I'll revert it back.

Reply:
I don't get it, why would a check of `ctx.substate` trigger parsing of the `.hgsubstate` file, if `submodule_mappings` is true, we will already have accessed it early in hg2git()? Please explain.

Reply:
For clarification:
Algorithm of updating list of submodules (when `.hgsubstate` file'd been added/modified) implemented as follows:
- removed all parents' subrepos
- added subrepos from current revision
---
If `.hgsubstate` file was emptied it can be recognized as modified (theoretically), but `ctx.substate` will be empty, and list of submodules won't be updated.
In fact I tested that case, and, wierdly, empty `.hgsubstate` had been considered as removed, but not modified. It might be related to "apology" comment about shortcut in `status` function, but for now I did't investigate it further. My point is: empty and changed substate doesn't mean that submodule list shouldn't be updated.

About overhead of parsing `.hgsubstate` file on `ctx.substate` call, this is a bit of ignorance from my side. I looked again in function `subrepoutil.state` and found out that parsed state is cached.

Reply:
If you think the argument is not sufficient, I will add substate check back (it shouldn't break anything anyway, I suppose)

Reply:
I buy your argument, good!